[bashible] fix bb-event-error-create

RomanenkoDenys opened the pull request:
Signed-off-by: Denis Romanenko <denis.romanenko@flant.com>

## Description
<!---
  Describe your changes in detail.

  Please let users know if your feature influences critical cluster components
  (restarts of ingress-controllers, control-plane, Prometheus, etc).
-->
Fix error in bb-event-error-create bashible function.

## Why do we need it, and what problem does it solve?
<!---
  This is the most important paragraph.
  You must describe the main goal of your feature.

  If it fixes an issue, place a link to the issue here.

  If it fixes an obvious bug, please tell users about the impact and effect of the problem.
-->
1. When bashible tries to create an event, we get an error because we do not have access from system:nodes group to events.k8s.io api.
2. If cluster in bootstrapping state, we cannot have working API to send events, so we simply suppress event creating errors to prevent bashible fail. 
## What is the expected result?
<!---
  How can one check these changes after applying?  

  Describe, what (resource, state, event, etc.) MUST or MUST NOT change/happen after applying these changes.
-->

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries
<!---
  Describe the changes so they will be included in a release changelog.

  Find examples and documentation below, or visit the instruction page on the repo wiki
  https://github.com/deckhouse/deckhouse/wiki/How-to-add-to-changelog
-->

```changes
section: candi
type: fix
summary: Fix error in bb-event-error-create bashible function.
impact_level: low
```

<!---
`impact_level: default` adds to changelog as usual, this is the default that can be omitted
`impact_level: high`    something important for users, the impact will be copied to "Know Before Update" section
`impact_level: low`     omitted in changelog YAML; note there is `type:chore` for chores

Tip for the section field:

  - <kebab-case of a module>, e.g. "cloud-provider-aws", "node-manager"
  - "ci", has forced low impact
  - "docs", includes website changes, should have low impact
  - "candi"
  - "deckhouse-controller"
  - "dhctl"
  - "global-hooks"
  - "go_lib"
  - "helm_lib"
  - "jq_lib"
  - "shell_lib"
  - "testing", has forced low impact
  - "tools", has forced low impact

Find changed sections:

gh pr diff   $PULL_REQUEST_NUMBER   |
  egrep "^([+]{3} b|[-]{3} a)/" |
  cut -d/ -f2- |
  sed 's#^ee/##' |
  sed 's#^fe/##' |
  sed 's#^modules/##' |
  sed 's#[0-9][0-9][0-9]-##' |
  egrep -v 'Makefile' |       # add file exclusion here
  cut -d/ -f1 |
  sort |
  uniq

Find all possible sections (excluding ci):

node -e 'console.log(require("./.github/scripts/js/changelog-find-sections.js")().join("\n"))'
-->

deckhouse-BOaTswain commented:
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->

yalosev commented on candi/bashible/bashible.sh.tpl:
```diff
@@ -27,10 +27,11 @@ function bb-event-error-create() {
     eventName="$(echo -n "$(hostname -s)")-$(echo $step | sed 's#.*/##; s/_/-/g')"
     eventLog="/var/lib/bashible/step.log"
-    kubectl apply -f - <<EOF
+    kubectl_exec apply -f - <<EOF || true
```
should not we also move `stderr` to `/dev/null` ?

RomanenkoDenys replied:
no because if kubectl exit code > 0 bashible fails

nabokihms commented on testing/matrix/linter/rules/roles/placement.go:
```diff
@@ -316,7 +316,7 @@ func handleNestedRBACForUs(m utils.Module, object storage.StoreObject, shortPath
 	switch {
 	case strings.HasPrefix(objectName, localPrefix):
-		if namespace != m.Namespace {
+		if namespace != m.Namespace && namespace != "default" {
```
Oh no 😞

nabokihms replied:
This condition must not be changed. `default` is considered a system namespace, so the rule's name must start with `d8:` prefix.

apolovov replied:
btw, it is better to use v1.NamespaceDefault.

RomanenkoDenys replied:
fixed by proper way

nabokihms commented on modules/040-node-manager/templates/bashible/rbac-for-us.yaml:
```diff
@@ -52,6 +52,20 @@ rules:
+    resources:
+      - events
+    verbs:
+      - "*"
```
Too broad, consider specifying only required verbs

RomanenkoDenys replied:
changed to get, list, create, update, patch

nabokihms commented on candi/bashible/bashible.sh.tpl:
```diff
@@ -41,7 +42,7 @@ function bb-event-error-create() {
           kind: Node
           name: '$(hostname -s)'
-          uid: "$(kubectl get node $(hostname -s) -o jsonpath='{.metadata.uid}')"
+          uid: "$(kubectl_exec get node $(hostname -s) -o jsonpath='{.metadata.uid}')"
```
BTW, uid is not required and can be avoided.

yalosev replied:
Nope. Without uid event is not binded to the object. I mean when you make `kubectl describe node X` - you will not fetch an event without id

nabokihms reviewed (changes requested):
Some questions regarding the event error create function:
1. The reason field is always `Failed`. `Failed` is too abstract. I expected to see an instruction on what should I do next, or at least, what is broken.

2. The action field is always equal to `Binding`. I assume it should be something like `BashibleFailed`.
> action is what action was taken/failed regarding to the regarding object. It is machine-readable. This field cannot be empty for new Events and it can have at most 128 characters.

What is the meaning of the `Binding`?

3. Why is there the `annotations.timestamp` field? It seems like eventTime is enough.

RomanenkoDenys commented:
> Some questions regarding the event error create function:
> 
> 1. The reason field is always `Failed`. `Failed` is too abstract. I expected to see an instruction on what should I do next, or at least, what is broken.

changed to BashibleStepFailed

> 2. The action field is always equal to `Binding`. I assume it should be something like `BashibleFailed`.
> 
> > action is what action was taken/failed regarding to the regarding object. It is machine-readable. This field cannot be empty for new Events and it can have at most 128 characters.
> 
> What is the meaning of the `Binding`?

changed to BashibleStepExecution

> 
> 3. Why is there the `annotations.timestamp` field? It seems like eventTime is enough.

removed

deckhouse-BOaTswain commented:
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...
<!-- workflow_start -->
